L1Trigger/L1THGCal/plugins/ntuples/HGCalTriggerNtupleGenTau.cc

```diff
@@ -0,0 +1,374 @@
+                        n_pi++;
+                        finalProd_p4.push_back(daughter->p4());
+                        finalProd_id.push_back(daughter->pdgId());
+                        tau_p4vis+=(daughter->p4());
+                    }                
+                    if( isNeutralPion( *daughter ) ){
+                        n_piZero++;
+                        const edm::RefVector<std::vector<reco::GenParticle> >& grandaughters = (*daughter).daughterRefVector();
+                        for( const auto grandaughter : grandaughters ){
+                            if( isGamma( *grandaughter ) ){
+                                n_gamma++;
+                                finalProd_p4.push_back(grandaughter->p4());
+                                finalProd_id.push_back(grandaughter->pdgId());
+                                tau_p4vis+=(grandaughter->p4());         
+                            }
+                        }
+                    }
+                }
+
+                /* Here the selection of the decay product according to the Pythia6 decayTree */
+                else if( !isPythia8generator_ ){            
+                    if( isChargedPion( *daughter ) ){
+                        n_pi++;
+                        finalProd_p4.push_back(daughter->p4());
+                        finalProd_id.push_back(daughter->pdgId());
+                        tau_p4vis+=(daughter->p4());
+                    }
+                    if( isNeutralPion( *daughter ) ){
+                        n_piZero++;
+                        const edm::RefVector<std::vector<reco::GenParticle> >& grandaughters = (*daughter).daughterRefVector();
+                        for( const auto grandaughter : grandaughters ){
+                            n_gamma++;
+                            finalProd_p4.push_back(grandaughter->p4());
+                            finalProd_id.push_back(grandaughter->pdgId());
+                            tau_p4vis+=(grandaughter->p4());         
+                        }          
+                    }
+                    if( isIntermediateResonance( *daughter ) ){
+                        const edm::RefVector<std::vector<reco::GenParticle> >& grandaughters = (*daughter).daughterRefVector();
+                        for( const auto grandaughter : grandaughters ){
+                            if( isChargedPion( *grandaughter ) ){
+                                n_pi++;
+                                finalProd_p4.push_back(grandaughter->p4());
+                                finalProd_id.push_back(grandaughter->pdgId());
+                                tau_p4vis+=(grandaughter->p4());         
+                            }                            
+                            if( isNeutralPion( *grandaughter ) ){
+                                n_piZero++;
+                                const edm::RefVector<std::vector<reco::GenParticle> >& descendants = (*grandaughter).daughterRefVector();
```

Comment:
Use `GenParticleRefVector` instead of `edm::RefVector<std::vector<reco::GenParticle> >`
Use `grandaughter->` instead of (*grandaughter).`